compiler/rustc_trait_selection/src/traits/effects.rs

```diff
@@ -232,6 +240,63 @@ fn evaluate_host_effect_from_item_bounds<'tcx>(
     }
 }
 
+/// Assembles constness bounds "normal" item bounds on aliases, which may include
+/// unconditionally `const` bounds that are *not* conditional and thus always hold.
+fn evaluate_host_effect_from_item_bounds<'tcx>(
+    selcx: &mut SelectionContext<'_, 'tcx>,
+    obligation: &HostEffectObligation<'tcx>,
+) -> Result<ThinVec<PredicateObligation<'tcx>>, EvaluationFailure> {
+    let infcx = selcx.infcx;
+    let tcx = infcx.tcx;
+    let drcx = DeepRejectCtxt::relate_rigid_rigid(selcx.tcx());
+    let mut candidate = None;
+
+    let mut consider_ty = obligation.predicate.self_ty();
+    while let ty::Alias(kind @ (ty::Projection | ty::Opaque), alias_ty) = *consider_ty.kind() {
+        for clause in tcx.item_bounds(alias_ty.def_id).iter_instantiated(tcx, alias_ty.args) {
+            let bound_clause = clause.kind();
+            let ty::ClauseKind::HostEffect(data) = bound_clause.skip_binder() else {
+                continue;
+            };
+            let data = bound_clause.rebind(data);
+            if data.skip_binder().trait_ref.def_id != obligation.predicate.trait_ref.def_id {
+                continue;
+            }
+
+            if !drcx.args_may_unify(
+                obligation.predicate.trait_ref.args,
+                data.skip_binder().trait_ref.args,
+            ) {
+                continue;
+            }
+
+            let is_match =
+                infcx.probe(|_| match_candidate(selcx, obligation, data, true, |_, _| {}).is_ok());
```

Comment:
```suggestion
                infcx.probe(|_| match_candidate(selcx, obligation, data, false, |_, _| {}).is_ok());
```

In trying to use this patch locally with my experimental const sizedness changes, I find that I get errors in `libcore` when this does normalisation.

I don't have a minimal example unfortunately, but the issue was when:

- `obligation.predicate.trait_ref` is `<<char as str::pattern::Pattern>::Searcher<'_> as marker::MetaSized>`
- `data.trait_ref` was `<<char as str::pattern::Pattern>::Searcher<'_> as marker::MetaSized>`
  - ..then normalised to `<str::pattern::CharSearcher<'_> as marker::MetaSized>`
- ..and then `is_match` ends up being `false`

I don't know that it's the *correct* solution but changing this to `false` allows the crate to compile successfully.

Reply:
Okay, I still need to work out some issues with the change and recursive structs, but will put a patch up after resolving that.

Reply:
#143391